sw/host/opentitanlib/src/util/bigint.rs

```diff
@@ -0,0 +1,320 @@
+            assert_eq!(TestArray::from_le_bytes(slice).unwrap().to_le_bytes(), data);
+        }
+        check(&[], &[0]);
+        check(&[1], &[1]);
+        check(&[0, 1], &[0, 1]);
+        check(&[1, 0], &[1]);
+
+        assert!(TestArray::from_le_bytes([1, 2, 3]).is_err());
+    }
+
+    #[test]
+    fn test_from_to_le_bytes_exact_len() {
+        fn check(slice: &[u8], data: &[u8]) {
+            assert_eq!(
+                TestArrayExact::from_le_bytes(slice).unwrap().to_le_bytes(),
+                data
+            );
+        }
+        check(&[0, 128], &[0, 128]);
+        check(&[255, 255, 0], &[255, 255]);
+
+        assert!(TestArrayExact::from_le_bytes([1]).is_err());
+        assert!(TestArrayExact::from_le_bytes([255, 127]).is_err());
+        assert!(TestArrayExact::from_le_bytes([0, 0, 1]).is_err());
+    }
+
+    #[test]
+    fn test_from_to_be_bytes() {
+        fn check(slice: &[u8], data: &[u8]) {
+            assert_eq!(TestArray::from_be_bytes(slice).unwrap().to_be_bytes(), data);
+        }
+        check(&[1], &[1]);
+        check(&[1, 0], &[1, 0]);
+        check(&[0, 1], &[1]);
+
+        assert!(TestArray::from_be_bytes([1, 2, 1]).is_err());
+    }
+
+    #[test]
+    fn test_from_to_be_bytes_exact_len() {
+        fn check(slice: &[u8], data: &[u8]) {
+            assert_eq!(
+                TestArrayExact::from_be_bytes(slice).unwrap().to_be_bytes(),
+                data
+            );
+        }
+        check(&[128, 1], &[128, 1]);
+        check(&[0, 255, 255], &[255, 255]);
+
+        assert!(TestArrayExact::from_be_bytes([1]).is_err());
+        assert!(TestArrayExact::from_be_bytes([127, 1]).is_err());
+        assert!(TestArrayExact::from_be_bytes([1, 0, 0]).is_err());
+    }
+
+    #[test]
+    fn test_bit_len() {
+        fn check(slice: &[u8], bit_len: usize) {
+            assert_eq!(TestArray::from_le_bytes(slice).unwrap().bit_len(), bit_len);
+        }
+        check(&[1], 1);
+        check(&[1, 0], 1);
+        check(&[255], 8);
+        check(&[0, 1], 9);
+        check(&[0, 128], 16);
+    }
+
+    #[test]
+    fn test_from_str() {
+        assert_eq!(TestArray::from_str("0x01").unwrap().to_le_bytes(), [1]);
+        assert_eq!(
+            TestArray::from_str("0x00201").unwrap().to_le_bytes(),
+            [1, 2]
+        );
+        assert!(TestArray::from_str("0x030201").is_err());
+    }
+
+    #[test]
+    fn test_from_str_exact_len() {
+        assert_eq!(
+            TestArrayExact::from_str("0x08001").unwrap().to_le_bytes(),
+            [1, 128]
+        );
+
+        assert!(TestArrayExact::from_str("0x01").is_err());
+        assert!(TestArrayExact::from_str("0x0201").is_err());
+        assert!(TestArrayExact::from_str("0x030201").is_err());
+    }
+
+    #[test]
+    fn test_fmt() {
+        let exact = TestArrayExact::from_str("0xabcd").unwrap();
+        assert_eq!(exact.to_string(), "0xabcd");
+        assert_eq!(format!("{}", exact), "0xabcd");
```

Comment:
Personally I don't think it's worth duplicating the `format!` version, since to_string() is defined by the standard library to work this way.